server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java

```diff
@@ -0,0 +1,162 @@
+
+import java.io.ByteArrayInputStream;
+import java.time.ZonedDateTime;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import javax.mail.MessagingException;
+
+import org.apache.james.core.User;
+import org.apache.james.mailbox.model.MailboxId;
+import org.apache.james.mailbox.model.MessageId;
+import org.apache.james.server.core.MailImpl;
+import org.apache.james.server.core.MimeMessageCopyOnWriteProxy;
+import org.apache.james.server.core.MimeMessageInputStreamSource;
+import org.apache.james.util.MemoizedSupplier;
+import org.apache.james.util.MimeMessageUtil;
+import org.apache.james.util.OptionalUtils;
+import org.apache.mailet.Attribute;
+import org.apache.mailet.AttributeName;
+import org.apache.mailet.AttributeValue;
+import org.apache.mailet.Mail;
+
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+
+public class MailAdapter {
+    private static final AttributeName ORIGIN_MAILBOXES_ATTRIBUTE_NAME = AttributeName.of("originMailboxes");
+    private static final AttributeName HAS_ATTACHMENT_ATTRIBUTE_NAME = AttributeName.of("hasAttachment");
+    private static final AttributeName OWNER_ATTRIBUTE_NAME = AttributeName.of("owner");
+    private static final AttributeName DELIVERY_DATE_ATTRIBUTE_NAME = AttributeName.of("deliveryDate");
+    private static final AttributeName DELETION_DATE_ATTRIBUTE_NAME = AttributeName.of("deletionDate");
+    private static final AttributeName SUBJECT_ATTRIBUTE_VALUE = AttributeName.of("subject");
+
+    private final MailboxId.Factory mailboxIdFactory;
+    private final MessageId.Factory messageIdFactory;
+
+    public MailAdapter(MailboxId.Factory mailboxIdFactory, MessageId.Factory messageIdFactory) {
+        this.mailboxIdFactory = mailboxIdFactory;
+        this.messageIdFactory = messageIdFactory;
+    }
+
+    Mail toMail(DeletedMessage deletedMessage) throws MessagingException {
+        MimeMessageCopyOnWriteProxy mimeMessage = new MimeMessageCopyOnWriteProxy(new MimeMessageInputStreamSource(deletedMessage.getMessageId().serialize(), deletedMessage.getContent().get()));
+
+        Collection<AttributeValue<?>> serializedMailboxIds = deletedMessage.getOriginMailboxes().stream()
+            .map(MailboxId::serialize)
+            .map(AttributeValue::of)
+            .collect(Guavate.toImmutableList());
+
+        AttributeValue<?> subjectAttributeValue = AttributeValue.of(deletedMessage.getSubject().map(AttributeValue::of));
+
+        return MailImpl.builder()
+            .name(deletedMessage.getMessageId().serialize())
+            .sender(deletedMessage.getSender())
+            .recipients(deletedMessage.getRecipients())
+            .mimeMessage(mimeMessage)
+            .attribute(new Attribute(OWNER_ATTRIBUTE_NAME, AttributeValue.of(new SerializableUser(deletedMessage.getOwner()))))
```

Comment:
couldn't we simplify this Attribute syntax? it looks so ugly

Reply:
I'd be glad to do it...